### PR TITLE
Hw 4

### PR DESCRIPTION
**Pull request by 3apy6a**

### CODE REVIEW

File: Rack/rack_app.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+require './config/initializers/initializer'
+require './run.rb'
+
+class RackApp
+  def call(env)
+    request = Rack::Request.new(env)
+    case request.path.start_with?
+    when '/api/sups'
+      Handlers::Sup(request).response
+    when ',api/customers'
```
Comment (Patriarh22) [inline Rack/rack_app.rb]: Are you sure you need a comma here?

File: Rack/rack_app.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+require './config/initializers/initializer'
+require './run.rb'
+
+class RackApp
+  def call(env)
+    request = Rack::Request.new(env)
+    case request.path.start_with?
+    when '/api/sups'
+      Handlers::Sup(request).response
```
Comment (Patriarh22) [inline Rack/rack_app.rb]: You call here the class method `response`, but I can't find this method in the list of class methods of `Handlers::Sup`, this class doesn't have class methods at all. But I can see the instance method `response`. Are you sure this logic works well?

File: Rack/rack_app.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    when ',api/customers'
+      Handlers::Customer(request).response
+    else
+      Handlers::Base.response(request)
```
Comment (Patriarh22) [inline Rack/rack_app.rb]: Class and instance of `Handlers::Base` doesn't have a method `response`

File: Rack/config/database.yml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+development:
```
Comment (Patriarh22) [inline Rack/config/database.yml]: It's not critical for now, but on a real project, you shouldn't add any credentials (to the database or other secrets) to the git repository, so I recommend you to unstage this file from this branch. You need this file locally, but not in the repository.

File: Rack/app/models/bookings_gear.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+class BookingsGear < ActiveRecord::Base
```
Comment (Patriarh22) [inline Rack/app/models/bookings_gear.rb]: This is a redundant model, you need it if you create `has_many through` relation. But if you use `has_and_belongs_to_many` relation, you must create only join table in the database, but without a model. 

Notice, you use `dependent: :destroy` in `belongs_to` relation, so if you destroy any `BookingsGear` record, you will destroy related `booking` and `gear` records too.

File: Rack/app/models/booking.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  validate :end_date_after_start_date
+
+  def booked_date_range
+    { from: :start_date, to: :end_date }
+  end
+
+  private
+
+  def end_date_after_start_date
+    errors.add(:end_date, 'must be after the start date') unless end_date < start_date
```
Comment (Patriarh22) [inline Rack/app/models/booking.rb]: you are adding error here when `end_date` is later than `start_date`

**Review by Patriarh22 — CHANGES_REQUESTED** [review]

File: Rack/rack_app.rb (hunk 1 above)
Comment (3apy6a) [reply Rack/rack_app.rb]: ups, removed it

File: Rack/rack_app.rb (hunk 2 above)
Comment (3apy6a) [reply Rack/rack_app.rb]: It's instance method.

File: Rack/app/models/booking.rb (hunk 6 above)
Comment (3apy6a) [reply Rack/app/models/booking.rb]: not a mandatory method, added simply, for filling.

File: Rack/rack_app.rb (hunk 3 above)
Comment (3apy6a) [reply Rack/rack_app.rb]: Not finished earlier

File: Rack/app/models/bookings_gear.rb (hunk 5 above)
Comment (3apy6a) [reply Rack/app/models/bookings_gear.rb]: understand

File: Rack/rack_app.rb [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  def call(env)
+    request = Rack::Request.new(env)
+
+    case request.path.start_with?
+    when '/api/sups'
+      Handlers::Sup.new(request).response
+    when 'api/customers'
+      Handlers::Customer.new(request).response
+    else
+      Rack::Response.new({ error: 'Not Found' }.to_json, 400, HEADERS).finish
```
Comment (Patriarh22) [inline Rack/rack_app.rb]: - Status `Not Found` has status code 404
- I think you can use here method `response_404` from `Handlers::Base`, there is the same logic

File: Rack/app/models/booking.rb (hunk 6 above)
Comment (Patriarh22) [reply Rack/app/models/booking.rb]: I agree with you, but it has to work correctly. The best way here it's using `if` instead of `unless`

**Review by Patriarh22 — CHANGES_REQUESTED** [review]

File: Rack/rack_app.rb (hunk 7 above)
Comment (3apy6a) [reply Rack/rack_app.rb]: OMG 400...

File: Rack/app/models/booking.rb (hunk 6 above)
Comment (3apy6a) [reply Rack/app/models/booking.rb]: why Rubocop writes : IF to long use UNLESS  bla bla bla?

Comment (Patriarh22) [reply Rack/app/models/booking.rb]: Rubocop wants that you should use `if` expression after `errors.add`, without wrapping around this action

File: Rack/rack_app.rb [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  def call(env)
+    request = Rack::Request.new(env)
+
+    case request.path.start_with?
+    when '/api/sups'
+      Handlers::Sup.new(request).response
+    when 'api/customers'
+      Handlers::Customer.new(request).response
+    else
+      Rack::Base.response_404
```
Comment (Patriarh22) [inline Rack/rack_app.rb]: I meant you should use `Handlers::Base.new.response_404`.
Cause `Rack::Base` will return you `uninitialized constant Rack::Base`

**Review by Patriarh22 — CHANGES_REQUESTED** [review]

File: Rack/rack_app.rb (hunk 8 above)
Comment (3apy6a) [reply Rack/rack_app.rb]: done

**Pull request merged**